gemini/use-cases/retrieval-augmented-generation/qna_using_query_routing/utils/qna_using_query_routing_utils.py

```diff
@@ -0,0 +1,163 @@
+    """
+    Loads documents from a folder and splits them into manageable chunks.
+
+    Supports both PDF and plain text documents.
+
+    Args:
+        index_path (str): The path to the folder containing the document(s).
+
+    Returns:
+        List['str']: A list of strings, representing the chunked document(s).
+    """
+
+    split_docs = []
+
+    if index_path[-1] != "/":
+        index_path = index_path + "/"
+    if index_path == "":
+        index_path = "."
+
+    for file_name in os.listdir(index_path):
+        print(f"Chunking input file: {file_name}")
+        if file_name.endswith(".pdf"):
+            loader = UnstructuredPDFLoader(index_path + file_name)
+        else:
+            loader = TextLoader(index_path + file_name)
+
+        text_splitter = CharacterTextSplitter(
+            chunk_size=int(config["vector_search"]["chunk_size"]),
+            chunk_overlap=int(config["vector_search"]["chunk_overlap"]),
+        )
+        split_docs.extend(text_splitter.split_documents(loader.load()))
+
+    return split_docs
+
+
+def create_vector_search_index(
+    bucket_uri: str,
+) -> tuple[aiplatform.MatchingEngineIndex, aiplatform.MatchingEngineIndexEndpoint, str]:
+    """
+    Creates a Vertex AI Matching Engine index, endpoint, and deploys the index.
+
+    Args:
+        bucket_uri (str): The Cloud Storage bucket URI where embedding data is stored.
+
+    Returns:
+        tuple: A tuple containing:
+            * aiplatform.MatchingEngineIndex: The created index object.
+            * aiplatform.MatchingEngineIndexEndpoint: The created index endpoint object.
+            * str: The ID of the deployed index.
+    """
+
+    print("Creating new vector search index..")
+
+    UID = datetime.now().strftime("%m%d%H%M")
+
+    # create index
+    my_index = aiplatform.MatchingEngineIndex.create_tree_ah_index(
+        display_name=config["vector_search"]["me_index_name"],
+        location=config["vector_search"]["me_region"],
+        contents_delta_uri=bucket_uri,
+        dimensions=int(config["vector_search"]["me_dimensions"]),
+        approximate_neighbors_count=int(
+            config["genai_qna"]["number_of_references_to_summarise"]
+        ),
+        distance_measure_type="DOT_PRODUCT_DISTANCE",
+    )
+    print(f"Created new index : {my_index.display_name} with ID: {my_index.name}")
+
+    # create IndexEndpoint
+    index_endpoint = aiplatform.MatchingEngineIndexEndpoint.create(
+        display_name=f'{config["vector_search"]["me_index_name"]}-endpoint',
+        public_endpoint_enabled=True,
+    )
+    print(
+        f"Created new index endpoint : {index_endpoint.display_name} with ID: {index_endpoint.name}"
+    )
+
+    deployed_index_id = (
+        f'{config["vector_search"]["me_index_name"].replace("-", "_")}_deployed_{UID}'
+    )
+
+    # deploy the Index to the Index Endpoint
+    index_endpoint.deploy_index(index=my_index, deployed_index_id=deployed_index_id)
```

Comment:
Note - This can take a very long time to run (about an hour). Make sure this function doesn't block other tasks if possible.